src/common.cu

```diff
@@ -80,6 +83,125 @@ static int cudaGraphLaunches = 0;
+  snprintf( columnsName, 1000, stdoutFormatName, "size", "count", "type", "redop", "root", timeStr, "algbw", "busbw", "#wrong", timeStr, "algbw", "busbw", "#wrong" );
+  std::cout << columnsName << std::endl;
+
+  if (persistanceMode){
+    const char csvFormatName[1000] {"size, count, type, redop, root, %s, algbw, busbw, #wrong, %s, algbw, busbw, #wrong, nthreads, ngpus, minbytes, maxbytes, stepbytes, stepfactor, check, warmup_iters, iters, agg_iters, op, datatype, root, parallel_init, blocking, stream_null, timeout, cudagraph, report_cputime\n"};
+    snprintf( columnsName, 1000, csvFormatName, timeStr, timeStr );
+    outf << columnsName;
+    outf.flush();
+  }
+
+  char columnsUnits[1000];
+
+  const char stdoutFormat[1000] { "# %10s  %12s  %8s  %6s  %6s  %7s  %6s  %6s  %5s  %7s  %6s  %6s  %5s" };
+  snprintf( columnsUnits, 1000, stdoutFormat, "(B)", "(elements)", "", "", "", "(us)", "(GB/s)", "(GB/s)", "", "(us)", "(GB/s)", "(GB/s)", "" );
+  std::cout << columnsUnits << std::endl;
+
+  if (persistanceMode){
+    outf << "(B), (elements),,,,(us),(GB/s),(GB/s),,(us),(GB/s),(GB/s),,,,,,,,,,,,,,,,,,,,\n";
+    outf.flush();
+  }
+}
+
+void Reporter::parameters(long bytes, long elements, const char* typeName, const char* opName, int rootName){
+  if (!isMainThread()){
+    return;
+  }
+
+  char parameters[1000];
+  
+  const char stdoutFormat[1000] { "%12li  %12li  %8s  %6s  %6i" };
+  snprintf( parameters, 1000, stdoutFormat, bytes, elements, typeName, opName, rootName );
+  std::cout << parameters;
+
+  if (persistanceMode){
+    const char csvFormat[1000] { "%li, %li, %s, %s, %i" };
+    snprintf( parameters, 1000, csvFormat, bytes, elements, typeName, opName, rootName );
+    outf << parameters;
+    outf.flush();
+  }
+}
+
+void Reporter::result(const char* time, float algBw, float busBw, long wrongElts){
+  if (!isMainThread()){
+    return;
+  }
+
+  char result[1000];
+
+  const char stdoutFormat[1000] { "  %7s  %6.2f  %6.2f  %5li" };
+  snprintf( result, 1000, stdoutFormat, time, algBw, busBw, wrongElts);
+  std::cout << result;
+
+  if (persistanceMode){
+    const char csvFormat[1000] { ", %s, %f, %f, %li" };
+    snprintf( result, 1000, csvFormat, time, algBw, busBw, wrongElts);
+    outf << result;
+    outf.flush();
+  }
+}
+
+void Reporter::newStep(){
+  if (!isMainThread()){
+    return;
+  }
+
+  std::cout << "\n";
+  if (persistanceMode){
+    const char csvFormat [1000] {",%i,%i,%lu,%lu,%lu,%lu,%i,%i,%i,%i,%s,%s,%i,%i,%i,%i,%i,%i,%i\n"};
+    char result[1000];
+    snprintf( result, 1000, csvFormat, nThreads, nGpus, minBytes, maxBytes, stepBytes, stepFactor, datacheck, warmup_iters, iters, agg_iters, test_opnames[ncclop], test_typenames[nccltype], ncclroot, parallel_init, blocking_coll, streamnull, timeout, cudaGraphLaunches, report_cputime);
+    outf << result;
+    outf.flush();
+  }
+}
+
+bool Reporter::isMainThread(){
+  return isMainThreadFlag;
+}
+
+Reporter::~Reporter(){
```

Comment:
Great to have created a `Reporter` unit. Clean and easy to reuse to export to another format.